include/wil/registry.h

```diff
@@ -17,16 +17,3139 @@
 
 #include <winreg.h>
 #include <new.h> // new(std::nothrow)
-#include "resource.h" // unique_hkey
+#include "registry_helpers.h"
+#include "resource.h"
+
+#ifndef __WIL_WINREG_
+#error This is required for ::wil::unique_hkey support
+#endif
+
+// wil registry does not require the use of the STL or C++ exceptions
+// wil registry natively supports std::vector and std::wstring when preferring those types
+// wil registry uses the __WIL_WINREG_STL define to enable support for wil::shared_* types (defined in resource.h)

```

Comment:
What does a user need to do to get this? Does this happen automatically since we include `resource.h` at the top of the file anyway? Maybe we don't need any `__WIL_WINREG_STL` guards then

Reply:
this goes back to how wil attempts to automatically enable functions based of what the caller is already using - it doesn't force new dependencies on the caller globally. it's a balancing act to avoid having a bazillion wil headers supporting different callers vs. callers including what they want before including wil headers.

It can be tricky in some cases for callers to get this right, but that's the current balancing act.

We don't get any of this for free - each of these __WIL__ definitions have their own prerequisites.
It sounds like the feedback is to recurse through each of these and detail more clearly to the user of registry.h what they need to have previously included, as specific as possible.

Reply:
Oh, I mean specifically how is `__WIL_WINREG_STL` controlled? This file includes `resource.h` already, so if it's defined there, then it's impossible to load this file without having `__WIL_WINREG_STL` defined.

Reply:
It looks like `__WIL_WINREG_STL` is also dependant on things like `WINAPI_FAMILY_PARTITION(WINAPI_PARTITION_DESKTOP) && !defined(WIL_KERNEL_MODE)`. So it's possible that it won't be undefined (which is unsurprising in retrospect; why would there be a redundant define?). Resolving.